sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/concatenate_data/concatenate_and_time_sync_node.hpp

```diff
@@ -82,115 +85,87 @@
+    std::unordered_map<std::string, double> & topic_to_original_stamp_map,
+    double reference_timestamp_min, double reference_timestamp_max);
 
 private:
-  /** \brief The output PointCloud publisher. */
-  rclcpp::Publisher<PointCloud2>::SharedPtr pub_output_;
-  /** \brief Delay Compensated PointCloud publisher*/
-  std::map<std::string, rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr>
-    transformed_raw_pc_publisher_map_;
-
-  /** \brief The maximum number of messages that we can store in the queue. */
-  int maximum_queue_size_ = 3;
-
-  double timeout_sec_ = 0.1;
-
-  bool publish_synchronized_pointcloud_;
-  bool keep_input_frame_in_synchronized_pointcloud_;
-  std::string synchronized_pointcloud_postfix_;
-
-  std::set<std::string> not_subscribed_topic_names_;
-
-  /** \brief A vector of subscriber. */
-  std::vector<rclcpp::Subscription<PointCloud2>::SharedPtr> filters_;
-
-  rclcpp::Subscription<geometry_msgs::msg::TwistWithCovarianceStamped>::SharedPtr sub_twist_;
-  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
-
-  rclcpp::TimerBase::SharedPtr timer_;
-  diagnostic_updater::Updater updater_{this};
-
-  const std::string input_twist_topic_type_;
-
-  /** \brief Output TF frame the concatenated points should be transformed to. */
-  std::string output_frame_;
-
-  /** \brief Input point cloud topics. */
-  // XmlRpc::XmlRpcValue input_topics_;
-  std::vector<std::string> input_topics_;
-
-  /** \brief TF listener object. */
-  std::shared_ptr<tf2_ros::Buffer> tf2_buffer_;
-  std::shared_ptr<tf2_ros::TransformListener> tf2_listener_;
-
-  std::deque<geometry_msgs::msg::TwistStamped::ConstSharedPtr> twist_ptr_queue_;
-
-  std::map<std::string, sensor_msgs::msg::PointCloud2::ConstSharedPtr> cloud_stdmap_;
-  std::map<std::string, sensor_msgs::msg::PointCloud2::ConstSharedPtr> cloud_stdmap_tmp_;
+  struct Parameters
+  {
+    int maximum_queue_size;
+    double timeout_sec;
+    bool is_motion_compensated;
+    bool publish_synchronized_pointcloud;
+    bool keep_input_frame_in_synchronized_pointcloud;
+    bool publish_previous_but_late_pointcloud;
+    std::string synchronized_pointcloud_postfix;
+    std::string input_twist_topic_type;
+    std::vector<std::string> input_topics;
+    std::string output_frame;
+    std::vector<double> lidar_timestamp_offsets;
+    std::vector<double> lidar_timestamp_noise_window;
+  } params_;
+
+  double current_concatenate_cloud_timestamp_{0.0};
+  double lastest_concatenate_cloud_timestamp_{0.0};
+  bool drop_previous_but_late_pointcloud_{false};
+  bool publish_pointcloud_{false};
+  double diagnostic_reference_timestamp_min_{0.0};
+  double diagnostic_reference_timestamp_max_{0.0};
+  std::unordered_map<std::string, double> diagnostic_topic_to_original_stamp_map_;
+
+  std::shared_ptr<CombineCloudHandler> combine_cloud_handler_;
+  std::shared_ptr<CloudCollector> cloud_collector_;
+  std::list<std::shared_ptr<CloudCollector>> cloud_collectors_;
```

Comment:
Having both is a bit confusing. Is `cloud_collector_` the currently active one? Please rename.

Reply:
`cloud_collector_` is the one that collects pointcloud in the same (similar) timestamp.
`cloud_collectors_` is a list that stores the multiple `cloud_collector_` for incoming pointcloud to compare their timestamp to, and if it matches the timestamp, the pointcloud will added to that pointcloud.

Reply:
I think calling `cloud_collector_` `current_cloud_collector_` or similar would be best :)

Reply:
@mojomex 
Sorry, this is a mistake, the `cloud_collector_` is not used.
Removed in https://github.com/autowarefoundation/autoware.universe/pull/8300/commits/130bcb87d6c697663a8bed2fd5de9f14290a39f9